akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -461,73 +544,68 @@ private[projection] class R2dbcOffsetStore(
           throw new IllegalArgumentException("Required EventEnvelope or DurableStateChange for TimestampOffset.")
         case _ =>
           throw new IllegalArgumentException(
-            "Mix of TimestampOffset and other offset type in same transaction isnot supported")
+            "Mix of TimestampOffset and other offset type in same transaction is not supported")
       }
-      saveTimestampOffsetInTx(conn, records)
+      saveTimestampOffsetInTx(conn, records, canBeConcurrent)
     } else {
       savePrimitiveOffsetInTx(conn, offsets.last.offset)
     }
   }
 
-  private def saveTimestampOffsetInTx(conn: Connection, records: immutable.IndexedSeq[Record]): Future[Done] = {
-    idle.set(false)
-    val oldState = state.get()
-    val filteredRecords = {
-      if (records.size <= 1)
-        records.filterNot(oldState.isDuplicate)
-      else {
-        // use last record for each pid
-        records
-          .groupBy(_.pid)
-          .valuesIterator
-          .collect {
-            case recordsByPid if !oldState.isDuplicate(recordsByPid.last) => recordsByPid.last
-          }
-          .toVector
+  private def saveTimestampOffsetInTx(
+      conn: Connection,
+      records: immutable.IndexedSeq[Record],
+      canBeConcurrent: Boolean): Future[Done] = {
+    load(records.map(_.pid)).flatMap { oldState =>
+      val filteredRecords = {
+        if (records.size <= 1)
+          records.filterNot(oldState.isDuplicate)
+        else {
+          // use last record for each pid
+          records
+            .groupBy(_.pid)
+            .valuesIterator
+            .collect {
+              case recordsByPid if !oldState.isDuplicate(recordsByPid.last) => recordsByPid.last
+            }
+            .toVector
+        }
       }
-    }
-    if (hasForeignOffsets() && records.nonEmpty) {
-      val latestTimestamp =
-        if (records.size == 1) records.head.timestamp
-        else records.maxBy(_.timestamp).timestamp
-      updateLatestSeen(latestTimestamp)
-    }
-    if (filteredRecords.isEmpty) {
-      FutureDone
-    } else {
-      val newState = oldState.add(filteredRecords)
-
-      // accumulate some more than the timeWindow before evicting, and at least 10% increase of size
-      // for testing keepNumberOfEntries = 0 is used
-      val evictThresholdReached =
-        if (settings.keepNumberOfEntries == 0) true else newState.size > (newState.sizeAfterEvict * 1.1).toInt
-      val evictedNewState =
-        if (newState.size > settings.keepNumberOfEntries && evictThresholdReached && newState.window
-              .compareTo(evictWindow) > 0) {
-          val evictUntil = newState.latestTimestamp.minus(settings.timeWindow)
```

Comment:
This could have been a problem after scaling, where latestTimestamp could be far ahead from received events from other slices.